zebra-scan/src/tests.rs

```diff
@@ -0,0 +1,371 @@
+    let nf = Nullifier([7; 32]);
+
+    let cb = fake_compact_block(
+        1u32.into(),
+        BlockHash([0; 32]),
+        nf,
+        &dfvk,
+        1,
+        false,
+        Some(0),
+    );
+
+    // The fake block function will have our transaction and a random one.
+    assert_eq!(cb.vtx.len(), 2);
+
+    let res = scan_block(
+        &zcash_primitives::consensus::MainNetwork,
+        cb.clone(),
+        &vks[..],
+        &[(account, nf)],
+        None,
+    )
+    .unwrap();
+
+    // The response should have one transaction relevant to the key we provided.
+    assert_eq!(res.transactions().len(), 1);
+    // The transaction should be the one we provided, second one in the block.
+    // (random transaction is added before ours in `fake_compact_block` function)
+    assert_eq!(res.transactions()[0].txid, cb.vtx[1].txid());
+    // The block hash of the response should be the same as the one provided.
+    assert_eq!(res.block_hash(), cb.hash());
+
+    Ok(())
+}
+
+/// Convert a zebra block and meta data into a compact block.
+fn block_to_compact(block: Arc<Block>, chain_metadata: ChainMetadata) -> CompactBlock {
+    CompactBlock {
+        height: block
+            .coinbase_height()
+            .expect("verified block should have a valid height")
+            .0
+            .into(),
+        hash: block.hash().bytes_in_display_order().to_vec(),
+        prev_hash: block
+            .header
+            .previous_block_hash
+            .bytes_in_display_order()
+            .to_vec(),
+        time: block
+            .header
+            .time
+            .timestamp()
+            .try_into()
+            .expect("unsigned 32-bit times should work until 2105"),
+        header: block
+            .header
+            .zcash_serialize_to_vec()
+            .expect("verified block should serialize"),
+        vtx: block
+            .transactions
+            .iter()
+            .cloned()
+            .enumerate()
+            .map(transaction_to_compact)
+            .collect(),
+        chain_metadata: Some(chain_metadata),
+
+        // The protocol version is used for the gRPC wire format, so it isn't needed here.
+        proto_version: 0,
+    }
+}
+
+/// Convert a zebra transaction into a compact transaction.
+fn transaction_to_compact((index, tx): (usize, Arc<Transaction>)) -> CompactTx {
+    CompactTx {
+        index: index
+            .try_into()
+            .expect("tx index in block should fit in u64"),
+        hash: tx.hash().bytes_in_display_order().to_vec(),
+
+        // `fee` is not checked by the `scan_block` function. It is allowed to be unset.
+        // <https://docs.rs/zcash_client_backend/latest/zcash_client_backend/proto/compact_formats/struct.CompactTx.html#structfield.fee>
+        fee: 0,
+
+        spends: tx
+            .sapling_nullifiers()
+            .map(|nf| CompactSaplingSpend {
+                nf: <[u8; 32]>::from(*nf).to_vec(),
+            })
```

Comment:
Optional: this would be easier to read and more reusable as a `Nullifier::to_bytes()` method in
https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/sapling/note/nullifiers.rs

Example:
https://github.com/ZcashFoundation/zebra/blob/2efc6bf4b2875f16a987bcbff0f90cff98eda3fa/zebra-chain/src/value_balance.rs#L334-L335

### Things we can't do

It would be nice to `impl From<sapling::Nullifier> for CompactSaplingSpend`, and `impl From<sapling::Output> for CompactSaplingOutput`, but that would move these dependencies to `zebra-chain`. (Due to Rust's orphan trait implementation rules.)

So we could do that behind a feature in `zebra-chain`, but it seems out of scope. Let's see if it's needed later.